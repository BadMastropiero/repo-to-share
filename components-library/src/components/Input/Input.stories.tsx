import {ComponentMeta, ComponentStory} from '@storybook/react';
import {Button} from '../Button';
import {Send} from '../Icon';

import Input from './Input';

export default {
  title: 'Components/Input',
  component: Input,
} as ComponentMeta<typeof Input>;

export const Base: ComponentStory<typeof Input> = args => {
  const {...rest} = args;

  return <Input {...rest} />;
};

Base.args = {
  disabled: false,
  text: 'Write here',
  required: false,
  type: 'text',
  primary: true,
};

Base.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/epHXucNHjke2rES69iOsaS/TetraImpacts?node-id=2%3A2838',
  },
};

export const WithButton: ComponentStory<typeof Input> = () => (
  <Input
    primary
    text="Your input with button"
    afterComponent={
      <Button
        style={{
          height: '36px',
          aspectRatio: 1,
        }}
        presetSize="normal"
        afterIcon={Send}
        primary
        rounded
      />
    }
  />
);

export const Phone: ComponentStory<typeof Input> = ({value, onChange}) => (
  <Input
    primary
    type="tel"
    text="Your phone number"
    maskPreset="phone"
    value={value}
    onChange={onChange}
  />
);

export const Email: ComponentStory<typeof Input> = ({value, onChange}) => (
  <Input
    type="tel"
    text="Your e-mail address"
    required
    maskPreset="email"
    value={value}
    onChange={onChange}
    primary
  />
);

export const Number: ComponentStory<typeof Input> = ({value, onChange}) => (
  <Input
    type="text"
    text="Your expected return"
    maskPreset="number"
    value={value}
    onChange={onChange}
    primary
  />
);
