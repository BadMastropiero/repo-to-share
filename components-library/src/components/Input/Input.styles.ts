import styled, {css} from 'styled-components';
import {Typography} from '../Typography';
import {flexAlignment} from '../../utils/styleMixins';

// Not in use, WIP
export const InputFieldset = styled.fieldset<{primary: boolean}>(
  ({
    theme: {
      colors: {primary, primary100},
      transition,
    },
  }) => css`
    transition: ${transition};
    position: absolute;
    inset: -5px 0px 0px;
    pointer-events: none;
    margin: 0;
    padding: 0;
    border: none;
    border-bottom: 4px dashed ${primary};
    &:disabled {
      border-bottom-color: ${primary100};
    }
  `
);

// Not in use, WIP
export const InputLegend = styled.legend<{primary: boolean}>(
  ({theme: {transition}}) => css`
    transition: ${transition};
    transform: scale(0.8);
    &.as-placeholder {
      //transform: translate(14px, 16px) scale(1);
      transform: translate(14px, 50%) scale(1);
    }
  `
);

export const InputLabel = styled(Typography)<{disabled?: boolean; primary?: boolean}>(
  ({
    theme: {
      colors: {primary, primary200, primaryDark, primaryBg},
      transition,
    },
    primary: isPrimary,
    disabled,
  }) => css`
    transition: ${transition};
    position: absolute;
    bottom: 100%;
    transform: translateY(50%);
    font-size: 11px;
    left: 0;
    pointer-events: none;

    color: ${() => {
      if (isPrimary) return disabled ? primary200 : primary;
      return disabled ? primaryDark : primaryBg;
    }};
    &.as-placeholder {
      bottom: 50%;
      font-size: 100%;
    }
  `
);

export const StyledInput = styled.input<{disabled?: boolean; $primary?: boolean}>(
  ({
    theme: {
      colors: {primary, caret, primaryBg, primary200, primaryDark},
      transition,
    },
    $primary: isPrimary,
    disabled,
  }) => css`
    transition: ${transition};
    pointer-events: all;

    -webkit-appearance: none;
    -moz-appearance: none;
    -o-appearance: none;
    appearance: none;

    width: 100%;
    padding: 0px;
    border-width: 0;

    font: inherit;
    caret-color: ${caret};
    color: ${() => {
      if (isPrimary) return disabled ? primary200 : primary;
      return disabled ? primaryDark : primaryBg;
    }};
    background-color: transparent;

    &:focus,
    &.noEmpty {
      outline: none;
    }
  `
);

export const InputWrapper = styled.div<{disabled?: boolean; primary?: boolean}>(
  ({
    theme: {
      colors: {primary, primaryBg, primary200, primaryDark},
      transition,
    },
    disabled,
    primary: isPrimary,
  }) => css`
    ${flexAlignment('row-center')}
    height: 52px;
    transition: ${transition};
    border-bottom: 4px dashed;
    border-color: ${() => {
      if (isPrimary) return disabled ? primary200 : primary;
      return disabled ? primaryDark : primaryBg;
    }};
  `
);

export const InputLabelWrapper = styled.div`
  position: relative;
  align-self: end;
  /* width: 100%; */
  flex: 1;
`;

export default StyledInput;
