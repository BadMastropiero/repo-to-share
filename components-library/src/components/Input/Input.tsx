import React, {useContext, useState, useEffect} from 'react';
import {ThemeContext} from 'styled-components';

// Components
import MaskedInput from 'react-text-mask';
import {Typography} from '../Typography';

// Styles
import {StyledInput, InputWrapper, InputLabel, InputLabelWrapper} from './Input.styles';

// Interfaces
import {InputProps} from './Input.interface';
import {Icon} from '../Icon';

const Input: React.FC<InputProps> = ({
  primary,
  disabled = false,
  text,
  value: valueProp,
  onChange,
  className,
  required,
  mask: maskProp = false,
  maskPreset,
  afterIcon,
  afterComponent,
  labelLineNumbers,
  ...rest
}) => {
  let mask = maskProp;
  const {input: inputTheme} = useContext(ThemeContext);
  if (inputTheme) {
    const maskPresetsTheme = inputTheme.maskPresets;
    if (maskPreset) mask = maskPresetsTheme[maskPreset] || false;
  }

  const [value, setValue] = useState(valueProp || '');
  const [focused, setFocused] = useState(false);
  const [hovered, setHovered] = useState(false);

  const onInputChange = (e: React.FormEvent<HTMLInputElement>) => {
    setValue((e.target as HTMLInputElement).value);
    if (onChange) onChange(e);
  };

  useEffect(() => {
    if (valueProp === value || !valueProp) return;
    setValue(valueProp);
  }, [valueProp]);

  return (
    <Typography style={{overflow: 'visible'}} fontSize="20px" lineHeight="200%" padding="0">
      <InputWrapper disabled={disabled} primary={primary} className="ti-input-wrapper">
        <InputLabelWrapper>
          <MaskedInput
            mask={mask}
            className={`${value || value !== '' ? 'noEmpty' : ''} ${className || ''}`}
            disabled={disabled}
            value={valueProp}
            onChange={onInputChange}
            onFocus={() => {
              setFocused(true);
            }}
            onMouseEnter={() => {
              setHovered(true);
            }}
            onMouseLeave={() => {
              setHovered(false);
            }}
            onBlur={() => {
              setFocused(false);
            }}
            required={required}
            {...rest}
            render={(ref, {defaultValue, ...restP}) => (
              <StyledInput
                as="input"
                ref={ref as (instance: HTMLInputElement | null) => void}
                $primary={primary}
                value={valueProp}
                defaultValue={!valueProp && defaultValue ? defaultValue : undefined}
                type="text"
                {...restP}
              />
            )}
          />
          <InputLabel
            lineNumbers={labelLineNumbers || 1}
            primary={primary}
            disabled={disabled}
            colorPreset="disable"
            variant="label"
            className={(!value || value === '') && !focused && !hovered ? 'as-placeholder' : ''}
          >
            {text} {required && '*'}
          </InputLabel>
        </InputLabelWrapper>
        {afterIcon && <Icon primary={primary} icon={afterIcon} />}
        {afterComponent}
      </InputWrapper>
    </Typography>
  );
};

export default Input;
