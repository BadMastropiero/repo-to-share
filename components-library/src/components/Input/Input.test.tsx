import {render, screen, fireEvent} from '@testing-library/react';
import {composeStories} from '@storybook/testing-react';
import '@testing-library/jest-dom';
import * as stories from './Input.stories';

// Every component that is returned maps 1:1 with the stories, but they already contain all decorators from story level, meta level and global level!
const {Base, Number, Phone} = composeStories(stories);

describe('Input tests', () => {
  test('should render', () => {
    render(<Base />);
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  test('should match snapshot', () => {
    const tree = render(<Base />);
    expect(tree).toMatchSnapshot();
  });

  it('should be controlled', () => {
    let value = '';
    const setValue = (number: string) => {
      value = number;
    };

    render(
      <Base
        value={value}
        onChange={e => {
          setValue((e.target as HTMLInputElement).value);
        }}
      />
    );
    const input = screen.getByRole('textbox');
    // userEvent.type(input, 'testing input'); // Not working on controlled input
    fireEvent.change(input, {target: {value: 'testing input'}});
    expect(value).toEqual('testing input');
  });
});

describe('Number Input tests', () => {
  it('should allow only numbers', () => {
    let value = '';
    const setValue = (number: string) => {
      value = number;
    };

    render(
      <Number
        value={value}
        onChange={e => {
          setValue((e.target as HTMLInputElement).value);
        }}
      />
    );
    const input = screen.getByRole('textbox');
    // userEvent.type(input, 'testing input 12 asd . 3'); // Not working on controlled input
    fireEvent.change(input, {target: {value: 'testing input 12 asd . 3'}});
    expect(value).toEqual('12.3');
  });
});

describe('Phone Button tests', () => {
  it('should match EEUU phone format', () => {
    let value = '';
    const setValue = (number: string) => {
      value = number;
    };

    render(
      <Phone
        value={value}
        onChange={e => {
          setValue((e.target as HTMLInputElement).value);
        }}
      />
    );
    const input = screen.getByRole('textbox');
    // userEvent.type(input, '(5asf65) 16#!#$5-16fwef wf5asfaf1sagg'); // Not working on controlled input
    fireEvent.change(input, {target: {value: '(5asf65) 16#!#$5-16fwef wf5asfaf1sagg'}});
    expect(value).toEqual('+(565) 165-1651');
  });
});
