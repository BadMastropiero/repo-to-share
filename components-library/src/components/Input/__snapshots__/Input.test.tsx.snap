// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bcXHqe sc-eDvSVe imORpF"
        font-size="20px"
        style="overflow: visible;"
      >
        <div
          class="sc-csuSiG hkTSoV ti-input-wrapper"
        >
          <div
            class="sc-eDWCr fNjHOp"
          >
            <input
              class="sc-ipEyDJ gvOjtI  "
              type="text"
              value=""
            />
            <label
              class="sc-bcXHqe sc-eDvSVe kMVOLM sc-kDvujY ivkgUA as-placeholder"
            >
              Write here
               
            </label>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bcXHqe sc-eDvSVe imORpF"
      font-size="20px"
      style="overflow: visible;"
    >
      <div
        class="sc-csuSiG hkTSoV ti-input-wrapper"
      >
        <div
          class="sc-eDWCr fNjHOp"
        >
          <input
            class="sc-ipEyDJ gvOjtI  "
            type="text"
            value=""
          />
          <label
            class="sc-bcXHqe sc-eDvSVe kMVOLM sc-kDvujY ivkgUA as-placeholder"
          >
            Write here
             
          </label>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
